### DISCUSSION TITLE
Model Options Support

### DISCUSSION THREAD

**Discussion by tzolov** [spring-projects/spring-ai#288] [Announcements]
With the new `ModelOptions` mechanism you can set any model-specific configurations on startup or at runtime.

For example: `seed`, `model`, `temperature`, `logitBias`, `topP`…  in case OpenAIChatClient or `num-thread`, `top-k`, `num-gpu` ... for the OllamaChatClient.

Use the auto-configuration `options` properties to set the default (e.g. start-up) options common for all model requests. 
Example for OpenAiChatClient:
```
spring.ai.openai.api-key=YOUR_API_KEY
spring.ai.openai.chat.options.model=gpt-35-turbo
spring.ai.openai.chat.options.temperature=0.7
```
or use the dedicated options builders, to create the options programmatically.

```java
var options = OpenAiChatOptions.builder()
        .withModel("gpt-35-turbo")
        .withTemperature(0.4)
        .withMaxTokens(200)
    .build()
  
var chatClient = new OpenAiChatClient(...).withDefaultOptions(options); // default options common for all request.

var prompt  = new Prompt(messages, options); // request-specific options that override the default such.

var response = chatClient.call(prompt);
```

NOTE: The runtime options override the default one.

For more information, check the updated [openai-chat](https://docs.spring.io/spring-ai/reference/api/clients/openai-chat.html) , and [openai-embeddings](https://docs.spring.io/spring-ai/reference/api/embeddings/openai-embeddings.html) documentations.

Gradually we will enable the options support to all model clients.
The Azure OpenAI Chat/Embeddings Options PR is underway: [#286](https://github.com/spring-projects/spring-ai/pull/286) and the OllamaChatClient has a partial support. 

Following diagram illustrates the data (and processing) flow for all model types:
![Spring AI (29)](https://github.com/spring-projects/spring-ai/assets/1351573/e04175da-302c-4f43-bc95-78444defdc76)

**Comment by tha2015** [comment]
I have some concerns regarding Options-related logic. Currently at startup time, a (default) Options is created, and at runtime, a (runtime) Options is created, then  some logic will build (native) Options from these two Options using different merge methods.
And the flow is like:
1. Create (native) Options
2. Copy values from (default) Options to (native) Options
3. Copy values from (runtime) Options to (native) Options

I believe that writing logic to merge is error prone especially when copying values from different  Options types (e.g. ChatOptionsImpl/PortableFunctionCallingOptions/AzureOpenAiChatOptions) to (native) Options (i.e. ChatCompletionsOptions). Also it is impossible to clear a default value at runtime (i.e. use a null value at runtime instead of a non-null default value).

Can we reconsider how to support portable options (without merging) to avoid these issues? Thanks.

**Comment by anupsajjan** [comment]
Hi  @tzolov . I was looking for a support to dynamically change API KEY at runtime through options. I did not find any relevant information related to this. Can you help me know if this is supported?

Right now if I add spring.ai.openai.api-key=YOUR_API_KEY in application.properties, its working, but if I want to set the API key at runtime, then I dont see any method options through OpenAIChatOptions .

Thanks!
<img width="663" alt="image" src="https://github.com/user-attachments/assets/31438bd5-37c2-46d8-baaf-d39c9ae6f1ef" />
